### PR TITLE
Ping packet ID list

### PR DESCRIPTION
**Pull request by Polo1K**
Allows the server to send a ping packet at a rate greater than the client reply time

### CODE REVIEW

**Comment by Janmm14** [comment]
https://github.com/SpigotMC/BungeeCord/pull/2409

**Comment by Janmm14** [comment]
Also if you apply a hard limit it'd probably be better to use a CircularFifo data structure based off an array instead of a LinkedList

**Comment by Polo1K** [comment]
> Also if you apply a hard limit it'd probably be better to use a CircularFifo data structure based off an array instead of a LinkedList

Swapped to using an EvictingQueue.

File: proxy/src/main/java/net/md_5/bungee/ServerConnection.java [hunk 1]
```diff
@@ -25,8 +28,7 @@
     @Getter
     private final boolean forgeServer = false;
     @Getter
-    @Setter
-    private long sentPingId = -1;
+    private final Queue<KeepAliveData> keepAliveHistory = EvictingQueue.create(1024);
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/ServerConnection.java]: EvictingQueue is presized with an array of this - 1024 seems a bit excessive.
Right now the existing code is essentially an EvictingQueue of size 1 and things are going (mostly ok), surely even eg 32 would be plenty.

**Comment by Janmm14** [comment]
The problem is if the evictingqueue gets full the server doesn't see the answers of his pings anymore. So with 32 it might work for a vanilla server, but if plugins decide they want to send more pings, we should have a reasonable buffer present, i think 512 would be enough. Maybe we could make the size editable wirh a property or so.

File: proxy/src/main/java/net/md_5/bungee/ServerConnection.java (hunk 1 above)
Comment (Polo1K) [reply proxy/src/main/java/net/md_5/bungee/ServerConnection.java]: If a server wants to send a ping packet every tick, I feel the client should be given the chance to not reply for up to 30 seconds, the default timeout. That's at least 600, but I feel we should be generous given that it doesn't take much memory to store this data.

**Comment by md-5** [comment]
> Also if you apply a hard limit it'd probably be better to use a CircularFifo data structure based off an array instead of a LinkedList

@Janmm14 well surely given the discussion above, then there is no difference using a CircularFifo vs a LinkedList (+ the LinkedList then doesn't have a large initial size)?

**Comment by Janmm14** [comment]
@md-5 a circularfifo puts less stress on the gc compared to a linkedlist. Maybe we can change ack to time-based evicting like in my old PR and use an ArrayDeque (never-full, self-resizing circular fifo)

**Comment by Janmm14** [comment]
?

**Pull request closed without merging**